### ISSUE TITLE
Example writing onto existing numpy buffer dies

### ISSUE THREAD

**Issue by adybbroe** [Kozea/cairocffi#84]
We can confirm the problem raised by stuaxo on Feb 19, 2015. Running the same code with numpy 1.11.1 and 1.9.1 gives the same error with cairocffi. However, if we switch to the cffi-from-buffer branch it gives another error.

Output from cairocffi (branch cffi-from-buffer):

```
ValueError                                Traceback (most recent call last)
/home/a000680/dev/cairocffi_example.py in <module>()
     32 data = numpy.zeros((width, height, 4), dtype=numpy.uint8)
     33 surface = cairo.ImageSurface.create_for_data(
---> 34     data, cairo.FORMAT_ARGB32, width, height)
     35 cr = cairo.Context(surface)
     36 

/home/a000680/usr/src/cairocffi/cairocffi/surfaces.py in create_for_data(cls, data, format, width, height, stride)
    665 
    666         """
--> 667         return cls(format, width, height, data, stride)
    668 
    669     @staticmethod

/home/a000680/usr/src/cairocffi/cairocffi/surfaces.py in __init__(self, format, width, height, data, stride)
    657             pointer = cairo.cairo_image_surface_create_for_data(
    658                 ffi.cast('char*', address), format, width, height, stride)
--> 659         Surface.__init__(self, pointer, target_keep_alive=data)
    660 
    661     @classmethod

/home/a000680/usr/src/cairocffi/cairocffi/surfaces.py in __init__(self, pointer, target_keep_alive)
    138         self._pointer = ffi.gc(pointer, cairo.cairo_surface_destroy)
    139         self._check_status()
--> 140         if target_keep_alive not in (None, ffi.NULL):
    141             keep_alive = KeepAlive(target_keep_alive)
    142             _check_status(cairo.cairo_surface_set_user_data(

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

If you on line 140 change to the following:

```
        import numpy as np
        if np.all(target_keep_alive) not in (None, ffi.NULL):
```

The code seems to run correctly.

Alternatively we saw that we could get the example to run also with the master branch provided you flatten the data (`data.ravel()` instead of `data` as the first argument to `create_for_data()`) and do the following fix in `surface.py` in the `ImageSurface` class:

```
    def __init__(self, format, width, height, data=None, stride=None, tka=True):
        if data is None:
            pointer = cairo.cairo_image_surface_create(format, width, height)
        else:
            if stride is None:
                stride = self.format_stride_for_width(format, width)
            address, length = from_buffer(data)
            if length < stride * height:
                raise ValueError('Got a %d bytes buffer, needs at least %d.'
                                 % (length, stride * height))
            pointer = cairo.cairo_image_surface_create_for_data(
                ffi.cast('char*', address), format, width, height, stride)
        Surface.__init__(self, pointer, target_keep_alive=tka)
```

**Comment by liZe** [comment]
See #51.

**Comment by stuaxo** [comment]
I wasn't keen on ```elif 'numpy.ndarray' in str(type(obj))``` in #51, so I've fixed it below - 

```python
def from_buffer(obj):
    """Return ``(pointer_address, length_in_bytes)`` for a buffer object."""
    if hasattr(obj, 'buffer_info'):
        # Looks like a array.array object.
        address, length = obj.buffer_info()
        return address, length * obj.itemsize
    if hasattr(obj, '__array_interface__):
        # Numpy buffers.
        return obj.__array_interface__['data'][0], sum(obj,shape)
    else:
        # Other buffers.
        # XXX Unfortunately ctypes.c_char.from_buffer
        # does not have length information,
        # and we’re not sure that len(obj) is measured in bytes.
        # (It’s not for array.array, though that is taken care of.)
        return ctypes.addressof(ctypes.c_char.from_buffer(obj)), len(obj)
```

**Comment by senyai** [comment]
Hi! I'd like to fix the issue. One thing I need to know first is the expected shape. `(width, height, channels)` looks unnatural to me, because width steps (number of channels)x(arr.itemsize) bytes, height steps stride in bytes, and channel steps arr.itemsize bytes. That's not in order. People expect IMAGE_ARR[0] to be a row, so, I's like to fix the order as `(height, width, channels)`. Then have `def ImageSurface.__init__(self, ` with all parameters default to None: `format=None, width=None, height=None, data=None, stride=None)`, because when `numpy` array is passed, all these parameters are known. Does anyone mind? @Kozea?

**Comment by senyai** [comment]
Okay, I have a plan:

1. Get numpy arrays with `matplotlib.image.imread` by opening grayscale, RGB and RGBA images.
2. Pass these arrays to cairocffi's `ImageSurface`.
3. Draw some primitives.
4. Use `matplotlib.image.imsave` to ensure that everything worked fine, except for RGB image, which is expected.

**Comment by stuaxo** [comment]
This all sounds good.

You can compare the output of matplotlib.image.imsave to ImageSurface.write_png, and make sure e.g. that the colours are correct.

Is this going to convert the input to premultiplied alpha, and if so - will there be a way to not do the conversion ?

Also wondering about channel order, ircc, internally cairo is bgra, whereas a caller might expect to to pass in rgba.

**Comment by xy-23** [comment]
I change this:
> return obj.__array_interface__['data'][0], sum(obj,shape)

into this:
`return ctypes.addressof(ctypes.c_char.from_buffer(obj)), reduce(lambda x, y: x*y , obj.shape)`

and change this:
```
    def __init__(self, pointer, target_keep_alive=None):
        self._pointer = ffi.gc(
            pointer, _keepref(cairo, cairo.cairo_surface_destroy))
        self._check_status()
        if target_keep_alive not in (None, ffi.NULL):
            keep_alive = KeepAlive(target_keep_alive)
            _check_status(cairo.cairo_surface_set_user_data(
                self._pointer, SURFACE_TARGET_KEY, *keep_alive.closure))
            keep_alive.save()
```
into this: (add `.any()`)
```
    def __init__(self, pointer, target_keep_alive=None):
        self._pointer = ffi.gc(
            pointer, _keepref(cairo, cairo.cairo_surface_destroy))
        self._check_status()
        if target_keep_alive.any() not in (None, ffi.NULL):
            keep_alive = KeepAlive(target_keep_alive)
            _check_status(cairo.cairo_surface_set_user_data(
                self._pointer, SURFACE_TARGET_KEY, *keep_alive.closure))
            keep_alive.save()
```
and it works successfully.
btw, don't forget `from functools import reduce`

**Comment by liZe** [comment]
Closed by #152.
